crates/primitives/oracle/src/lib.rs

```diff
@@ -0,0 +1,63 @@
+use anyhow::bail;
+use serde::{Deserialize, Serialize};
+
+mod pragma;
+
+use pragma::*;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(tag = "oracle_name", content = "config")]
+pub enum Oracle {
+    Pragma(PragmaOracle),
+}
+
+impl Oracle {
+    pub fn new(oracle_name: &str, url: String, key: String) -> anyhow::Result<Self> {
+        match oracle_name {
+            "Pragma" => Ok(Oracle::Pragma(PragmaOracle::new(url, key))),
+            _ => bail!("Unknown Oracle name"),
+        }
+    }
+
+    pub fn set_base_url(&mut self, url: String) {
```

Comment:
```suggestion
    pub fn set_base_url(&mut self, url: String) -> &Self {
```
allows the chaining of constructors by returning a reference to the current instance